// ngx-tags-input/package/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var tags_input_component_1 = require("./src/tags-input.component");
var typeahead_1 = require("ngx-bootstrap/typeahead");
var forms_1 = require("@angular/forms");
var TagsInputModule = (function () {
    function TagsInputModule() {
    }
    TagsInputModule.forRoot = function () {
        return {
            ngModule: TagsInputModule,
            providers: []
        };
    };
    TagsInputModule.decorators = [
        { type: core_1.NgModule, args: [{
                    imports: [
                        common_1.CommonModule,
                        forms_1.FormsModule,
                        typeahead_1.TypeaheadModule.forRoot()
                    ],
                    declarations: [
                        tags_input_component_1.TagsInputComponent
                    ],
                    exports: [
                        tags_input_component_1.TagsInputComponent
                    ]
                },] },
    ];
    /** @nocollapse */
    TagsInputModule.ctorParameters = function () { return []; };
    return TagsInputModule;
}());
exports.TagsInputModule = TagsInputModule;
